library/file.lisp

```
@@ -0,0 +1,876 @@
+    (do
+     (p <- (tryinto-pathname path))
+     (pure (lisp Unit (p)
+             (cl:ensure-directories-exist p)
+             Unit))))
+
+  (declare directory-files ((TryintoPathname :a) => :a -> (Result FileError (List Pathname))))
+  (define (directory-files path)
+    "Returns all files within the directory. Returns an error if the pathname does not represent a directory."
+    (do
+     (p <- (tryinto-pathname path))
+     (dir <- (directory? p))
+      (if dir
+          (Ok (lisp (List Pathname) (p)
+                (uiop:directory-files p)))
+          (Err (PathError "Pathname does not represent a directory. Please add a trailing `\`." p)))))
+
+  (declare subdirectories ((TryintoPathname :a) => :a -> (Result FileError (List Pathname))))
+  (define (subdirectories path)
+    "Returns all subdirectories within the directory. Returns an error if the pathname does not represent a directory."
+    (do
+     (p <- (tryinto-pathname path))
+     (dir <- (directory? p))
+      (if dir
+          (Ok (lisp (List Pathname) (path)
+                (uiop:subdirectories path)))
+          (Err (PathError "Pathname does not represent a directory. Please add a trailing `\`." p)))))
+
+  ;;
+  ;; Handling directory behavior that depends on emptiness
+  ;;
+  
+  (declare empty? ((TryintoPathname :a) => :a -> (Result FileError Boolean)))
+  (define (empty? path)
+    "Checks whether a directory is empty."
+    (do
+     (p <- (tryinto-pathname path))
+     (dir <- (directory? p))
+      (if dir
+          (Ok (and (list:null? (unwrap (directory-files p)))
+                   (list:null? (unwrap (subdirectories p)))))
+          (Err (PathError "Pathname does not represent a directory" p)))))
+
+  (declare remove-directory ((TryintoPathname :a) => :a -> (Result FileError Unit)))
+  (define (remove-directory path)
+    "Deletes a directory `dir` if it's empty, otherwise, returns`FileError`."
+    (do
+     (p <- (tryinto-pathname path))
+     (mt <- (empty? p))
```

Don't check this, let it fail.